### ISSUE TITLE
(How) does SCTP handle CPU-bound congestion on JavaScript thread?

### ISSUE THREAD

**Issue by henbos** [w3c/webrtc-pc#2086]
How is the JavaScript application supposed to be able to apply back pressure?
CC @marinaciocea 

The spec says:

> When an RTCDataChannel message has been received via the underlying data transport with type type and data rawData, the user agent MUST queue a task to run the following steps:
>
> [Convert data to the appropriate JS object and fire "onmessage".]

I don't see any feedback mechanism to tell the underlying data transport that we're actually done consuming the message. It looks like the underlying transport is free to queue tasks ad-infinitum to fire messages. The message is just taken out of the SCTP buffer and put on a PostTask to another thread. So SCTP should be free to receive more messages, whether or not JavaScript can keep up.

What if that other thread (JavaScript) is really slow at processing "onmessage"? E.g. I'm being sent 10000 numbers p*q per second and I spend 1 second per message calculating the values of p and q. Will the task queue get overflooded?

Am I missing something?

**Comment by aboba** [comment]
This is a design problem with the eventing model of WebSockets/RTCDataChannel.  A better approach (if events were used at all) would have been to event the availability of data, then let the application retrieve the data all at once, as a sequence of messages.  That way the buildup of received (and unprocessed) data would apply backpressure.  Plus the transfer of data could avoid copies.

**Comment by henbos** [comment]
It sounds like it would be easy to address this by introducing a mode to say "don't auto-empty the buffer" and a method "popMessage()".

**Comment by lgrahl** [comment]
This could be a slight improvement for WebSockets but it would not work for SCTP-based data channels. SCTP has only one receive buffer per association. Thus, you would risk that one data channel *filling up* (for example a low priority file transfer or, even worse, a channel that is being completely ignored and never bound by the application) blocks receiving messages from another data channel (for example real-time sensor data).
Furthermore, doing this on a message-based level is insufficient since messages can be of arbitrary size (although Chrome refuses to accept that messages can be larger than 64 KiB :stuck_out_tongue:). If one side sends a 10 MiB message then you still have to buffer 10 MiB. The same applies to 100 MiB, and so on. That breaks the backpressure mechanism.

So, we not only need an API that allows to receive parts of a message (#1732) but also find a solution for the *one receive buffer per association* problem which requires some protocol work. I have some ideas how it could be done by using PPIDs to signal backpressure.

**Comment by murillo128** [comment]
I agree with @lgrahl that a per-message api is needed, so we are able to receive messages of and arbitrary size. and avoid buffering of the sctp chunks before delivering it to the app. 

Not sure if I agree on the one receive buffer per association, @lgrahl can you elaborate? Shouldn't interleave mode solve this?

**Comment by murillo128** [comment]
BTW regarding backpressure, couldn't this be done by the app itself sending
back an "ack" messages in another datachannel? Don't think we need an API
and protocol change for that.

**Comment by aboba** [comment]
Per data channel flow control won't necessarily solve the problem if there is also association flow control. We are learning this in QUIC where flow control of a file transfer on one stream can impact  real-time data on another stream (or potentially DATAGRAM traffic as well). If the application can't keep up on one stream then that stream can be flow controlled but this also may consume much of the per-association buffer limit, resulting in flow control on other streams even if their per-stream limit is not reached.  But the alternative - filling the event queue with file transfer messages - also affects the timeliness of delivery of real time data, and potentially in a worse manner if there is no back pressure on the file transfer.

**Comment by murillo128** [comment]
wouldn't that be a different issue?  SCTP already has a congestion control
(better or worse) that provides back pressure on sender side, causing
buffering on sender.

**Comment by aboba** [comment]
@murillo128 @lgrahl Was just pointing out that per-data channel flow control will not completely solve the problem of side-effects from an application not keeping up on another data channel. Today implementations don't limit the message events in the event queue, and the message is considered "consumed" once it is put into the event queue, so flow control does not kick in.

**Comment by murillo128** [comment]
So it is an implementation problem, not a spec problem.

SCTP allows to signal back the `a_rwnd` in tha `SACK` feedback message:
```
   Each SACK an endpoint receives contains an a_rwnd value.  This value
   represents the amount of buffer space the data receiver, at the time
   of transmitting the SACK, has left of its total receive buffer space
   (as specified in the INIT/INIT ACK).  Using a_rwnd, Cumulative TSN
   Ack, and Gap Ack Blocks, the data sender can develop a representation
   of the peer's receive buffer space.

```
and

```
      As DATA chunks are delivered to the ULP and released from the
      receive buffers, increment a_rwnd by the number of bytes delivered
      to the upper layer.  This is, in effect, opening up rwnd on the
      data sender and allowing it to send more data.  The data receiver
      SHOULD NOT increment a_rwnd unless it has released bytes from its
      receive buffer.  For example, if the receiver is holding
      fragmented DATA chunks in a reassembly queue, it should not
      increment a_rwnd.
```
The issue is that we are considering `ULP` to be the datachannel protocol, and not the js application. If the `a_rwnd` is not decremented when the data chunk are delivered to the datachannel layer for reassembling, but when the event is actually deliver to the js application, we have the back-pressure mechanism that will solve the issue you are describing.

This however imposes a hard limit of 4GB of the amount of buffer that can be "buffered" by the browser (and on individual message length). I think that is reasonable, as to be able to handle messages bigger than that we should allow a more fine grained API to be able to deliver partial data to the js layer of each message as the sctp data chunks are received (maybe with a bit of buffering for ordering).

**Comment by henbos** [comment]
The spec is not explicit about when the receive buffer is emptied, but there is no syncing with the JavaScript thread unless we make that explicit, so I do believe this is more than an implementation problem. Could the spec say something like this instead?

> When an RTCDataChannel message has been received via the underlying data transport with type type and data rawData, the user agent MUST queue a task to run the following steps (note: at this point, the receive buffer of the underlying data channel has not been decreased for rawData):
> - In parallel...
>   - Decrement the receive buffer of the underlying data transport for the size of rawData.
> - [Convert data to the appropriate JS object and fire "onmessage" (same steps as before).]

**Comment by tuexen** [comment]
Most implementations right now have some buffer space within the JS layer. When SCTP passes user data to its upper layer (the JS layer) it is not buffered by SCTP anymore and therefore the buffer space is considered freed. You would have to avoid the buffering in the JS layer...

**Comment by murillo128** [comment]
@henbos if we need to make it explicit, something like that would work for me. We could also specify (in the IETF datachannel draft maybe) that the `ULP` of the sctp layer is not the datachannel layer, but the javascript app.

**Comment by lgrahl** [comment]
> Shouldn't interleave mode solve this?

It doesn't add per-stream send/receive buffers.

> This however imposes a hard limit of 4GB of the amount of buffer that can be "buffered" by the browser (and on individual message length).

IIRC that 4 GiB hard limit would also be shared across all data channels, so the actual limit is not predictable. Also, this is hypothetical since all browsers use usrsctp which uses a fixed size buffer (which obviously is much lower than 4 GiB). As @tuexen explained, you either take the data out and usrsctp considers the buffer space free'd or you don't and the flow control gets you which leads to the problem of one channel blocking another. Sure, it's an implementation issue... but even if one has a signalling mechanism for the receive window and it would follow the mechanism as described by @henbos, if that implementation uses a fixed receive buffer for the SCTP association it would come to a halt in case one received message surpasses that buffer's size.

**Comment by lgrahl** [comment]
@aboba If I understand you correctly then what you're saying is that for an ideal world we'd need both: A flow control that doesn't ignore the JS application **and** a per-channel flow control. Since we already have a priority mechanism with SCTP ndata, I'd agree.

**Comment by aboba** [comment]
@lgrahl Yes. To avoid "buffer bloat", implementations probably should limit buffering of received data in the message queue on a per-data channel basis. If implemented well, ndata could probably help avoid starvation in situations where an application can't keep up on one data channel.

**Comment by murillo128** [comment]
>IIRC that 4 GiB hard limit would also be shared across all data channels, so the actual limit per message is not predictable. Also, this is hypothetical since all browsers use usrsctp which uses a fixed size buffer (which obviously is much lower than 4 GiB). 

Yes,  ot is shared between all the datachannels, and actual limits will be much lower,  so spec wise it is good.  To increase the actual message limits wr would need to change the API to deliver individual chunks of a datachannel message to the app as soon as they are received and dont buffer it. 

> As @tuexen explained, you either take the data out and usrsctp considers the buffer space free'd or you don't and the flow control gets you which leads to the problem of one channel blocking another. Sure, it's an implementation issue... but even if one had a signalling mechanism for the receive window and it would follow the mechanism as described by @henbos, if that implementation uses a fixed receive buffer for the SCTP association it would come to a halt in case one incoming message surpasses that buffer's size.

that's an implementation issue, either modify usersctp to support it or use anothet datachannel implementation.

**Comment by lgrahl** [comment]
> that's an implementation issue

I said that. :slightly_smiling_face: However, ...

> either modify usersctp to support it or use anothet datachannel implementation.

But you do realise that this is pretty unrealistic? Announcing a 4 GiB receive window is essentially lying if the buffer is dynamic. And it is even more unrealistic in the kernel.

**Comment by lgrahl** [comment]
> an API for delivering data to the app as soon as it is received on the sctp chunks, ndata for interleaving and the change to the a_rnwd handling would solve all the issues.

Even if we had all of that I'd disagree. The receive buffer is message and stream agnostic. As long as streams are sharing this buffer, one stream not being read from would eventually bring all other streams to a standstill.

We can happily discuss that further but I think we're slowly drifting off-topic. IIRC the original question of @henbos kind of went into the direction whether there is an easy fix for the current 1.0 API... and at this point I don't think there is one.

**Comment by murillo128** [comment]
The solution proposed by @henbos would work api wise for the given problem.

**Comment by lgrahl** [comment]
Please explain how it would prevent the receive buffer from filling up when using multiple channels.

**Comment by murillo128** [comment]
the issue is not about preventing the receive buffers to fill up,  but to
prevent the datachannel stack to enque events in the js loop faster than
they can be dispatched to the js app.

If the a_rcwnd is not freed until the event is actually delivered to the js
app, it will make the sctp buffer to fill up and the sender will not push
more data. Once the data is delivered, the new a_rcwnd will be transmitted
back to the sender and it will start sending data again.

That way the server will pace its sending to the amount of data that can be
consumed by the js app.

If it is from a single datachannel or multiple ones is an orthogonal issue.

**Comment by henbos** [comment]
Whenever there are multiple streams of traffic on the "same wire" there will be competition, some packets will get dropped and some packets will arrive. @lgrahl Why do you think that if the buffer is shared one data channel would end up dominating it? (How does SCTP prioritize different channels?) Is your concern in general use cases, or is this only a concern in edge cases such as if one data channel is sending messages the size of the entire buffer, leaving no room for other messages, and SCTP prioritizing the same channel every time? (Follow-up question: If there is a legitimate concern, is there no good way for the application to get around it?)

The fix I'm proposing does not change how SCTP work but it does provide a feedback mechanism between the JS thread and the network thread. JS will still have "onmessage" fire in accordance to SCTP receiving messages, it will however ensure that the buffer does not get emptied faster than it is processed by the application.

**Comment by murillo128** [comment]
the issue that @lgrahl is highlighting is an original one,  IMO. With
current api,  data is buffered by the dc stack until all the chuncks are
received.

So if ndata is in use, it may happen that non completed messages may drain
the receive buffer, stalling the sender.

But, again, that's a different problem that requires a different solution.
Maybe a new issue should be opened for not mixing both discussions?

**Comment by henbos** [comment]
Are these the two issues or am I confused?
1. (This issue) How to get SCTP congestion to kick in when you are CPU bound on the JavaScript thread (with [proposal earlier](https://github.com/w3c/webrtc-pc/issues/2086#issuecomment-457726283)).
2. (Unrelated to this issue) How to prioritize data channels when the send-side buffer is full so that one data channel does not end up dominating the send buffer. E.g. one channel is sending lots of tiny messages so that the buffer is always full, making it impossible for another channel to send its large messages because the buffer never has sufficient space.

**Comment by murillo128** [comment]
they're two different issues IMO too

**Comment by henbos** [comment]
We can file separate issues in that case, I just wanted to make sure we're on the same page about what the issues being discussed are. Did I get 2) right @lgrahl ?

**Comment by lgrahl** [comment]
The issue is that there are N potential streams sharing 1 single receive buffer. Let the receive buffer/window *W* be 1 MiB. Let there be streams *A* and *B* where *B* is, for some reason, not being read from. The sender sends 1 MiB to stream *B* and then to stream *A*. Also, terminology here is that it's a *stream* on SCTP level and a *channel* on JS level.

With the current API, this would happen:

1. Chunks of size S for stream B would be read from W into a separate reassembly buffer allocated by the browser for channel B. This frees the amount of data from W.
2. All chunks of the message have been received, the message is being reassembled and bubbled via `onmessage` on channel B.
3. No one listens on B but it doesn't matter.
4. The same happens for stream/channel A.

With the proposed change, this would happen:

1. Chunks of size S for stream B would be received and
  a) either copied from W into a separate reassembly buffer allocated by the browser for channel B, or
  b) just left in W until the message is complete.
  In both cases this does not free the amount of data from W.
2. All chunks of the message have been received, the message is being reassembled and bubbled via `onmessage` on channel B.
3. No one listens on B, thus W is now clogged as it only has a capacity of 1 MiB.
3. The message on stream/channel A cannot be received.

---

Same example, but now B is being read from and we send 2 MiB sized messages instead of just 1 MiB.

With the current API, nothing changes compared to the above example.

With the proposed change, this would happen:

1. Chunks of size S for stream B would be received and
  a) either copied from W into a separate reassembly buffer allocated by the browser for channel B, or
  b) just left in W until the message is complete.
  In both cases this does not free the amount of data from W.
2. This continues until W has reached its capacity (1 MiB). But the message is of size 2 MiB, so W is clogged and the association stalls.

---

@henbos I'm trying to make a point that *one receive buffer per association* is a problem for 1). I haven't really talked about the sender side 2) but IIRC ndata does indeed solve the priorisation issue (by the use of stream schedulers), so I don't think we need to file an issue for that. It may have gotten a little confusing with all the discussion if it would make a difference with a different API, ndata & co.

**Comment by murillo128** [comment]
can you open a different issue for that?  that is definitively a different
one

**Comment by lgrahl** [comment]
Then please, both of you, tell me what you believe this issue is about:

1. Feeding back a signal to the SCTP stack to let the flow control kick in and stop receiving data due to CPU load on the JS thread?
2. Feeding back a signal to whatever thread is handling the SCTP stack to stop delivering messages to the JS thread? (No, this does not allow the flow control to kick in or it would be equivalent to 1.)

**Comment by murillo128** [comment]
the issue is sending back a signal to the sender so it doesn't send more
data until the event queue has been processed and the data delivered to the
js app.

**Comment by lgrahl** [comment]
Then we're talking about the same thing. Please, explain where my example is incorrect.

**Comment by murillo128** [comment]
your issue is a valid one but it's different than the current one even If
the solutions may be similar,  that's why we propose you to open a
different issue to solve it.

**Comment by murillo128** [comment]
let's put it in a different view. Taking into account just one single
datachannel will the proposed solution solve the issue?

Also,  does the problem about multichannel buffering still exist without
taking into account the event loop queueing issue?

you may have a solution for not problems simultaneously,  but it also
requires a change in w3c apis and ietf specs.

To solve this issue we only need one line in the spec without api change.

**Comment by henbos** [comment]
> As long as streams are sharing this buffer, one stream not being read from would eventually bring all other streams to a standstill.

As I understand it, it is impossible not to read from a stream under normal conditions. As soon as the message has arrived, _W_ is emptied and "onmessage" fires (or with my proposal: _W_ is emptied when JS thread is no longer blocked). It doesn't matter if anybody listens to this event or not, from SCTP's point of view the message has been delivered. The only way a stream would not be read from is if it never arrives in its entirety, a half-received message clogging up _W_ forever. If messages are sent sequentially, I believe this is an impossibility (_B_'s message would complete before _A_'s messages are put in _W_). If messages are sent "in parallel" (doesn't have to be true parallelism, but assuming there is competition here), there may be a possibility of lots of small messages on one channel "choking" a large message on another channel that never fully completes in _W_. If so this could potentially be a problem for both the the receiver's receive buffer (discussed here) and sender's send buffer (analogous problem). I do believe this is a very important but separate problem from:

> Feeding back a signal to the SCTP stack to let the flow control kick in and stop receiving data due to CPU load on the JS thread?

As for this:

> Feeding back a signal to whatever thread is handling the SCTP stack to stop delivering messages to the JS thread? (No, this does not allow the flow control to kick in or it would be equivalent to 1.)

If congestion control kicks in then a consequence of that would be that messages are not delivered on the JS thread, because they have not fully arrived on _W_.

**Comment by henbos** [comment]
As for receiving messages larger than the entire buffer, I think that is a third issue, and would require copying chunks from the SCTP buffer to a pre-delivery buffer, which I'm not sure is an "implementation detail" or something that needs to be specified.

**Comment by murillo128** [comment]
I agree with @henbos there are already 3 different issues here, which should be easier to track in different issues and not all at once.

**Comment by murillo128** [comment]
> > To solve this issue we only need one line in the spec without api change.

>Unfortunately not.

I mean strictly solve this issue, i.e. preventing a the DC stack to push events to the event queue faster than what the main js queue can drain them.

The other issues are obviously not solved.

**Comment by murillo128** [comment]
Let's try to recap, theese are the three issues:

- With current API model, the DC stack can push event faster to the event queue that what the main js thread is able to handle, so event queue grows indefinetivelly.
- If interleaving is used, it could be possible that an large, low priority message fills the sctp buffer halting the sender and higher priority but smaller messages can't be received.
- It is possible that a message bigger than the receive buffer fills the sctp buffer, draining the halting the sender.

Can we agree on this?

**Comment by murillo128** [comment]
Also, shouldn't 3) be solved by correctly negotiating the `max-message-size`  in the SDP O/A ?

https://tools.ietf.org/html/draft-ietf-mmusic-sctp-sdp-26#page-8

**Comment by lgrahl** [comment]
@henbos 

> As I understand it, it is impossible not to read from a stream under normal conditions. As soon as the message has arrived [...]

Let's stop right there. With the proposed change, the message (of size M) will never arrive if M is greater W because of...

> [...] at this point, the receive buffer of the underlying data channel has not been decreased for rawData [...]

A suitable comparison would be a TCP socket with a receive buffer W where one is only peeking (!) for data of size M. This will stall the TCP connection if M > W.

@murillo128 

> [...] 1), 2) and 3). Can we agree on this?

1) is indeed what this was originally about but **the proposed change** would result in 2) and 3) becoming a problem in the first place. 2) and 3) would never happen with the current specification.

> Also, shouldn't 3) be solved by correctly negotiating the max-message-size in the SDP O/A ?

The receive buffer has nothing to do and should never impact the maximum message size (important: **it would** with the proposed change here which is why I'm still arguing!). For example, Firefox uses a buffer of 1 MiB but can receive messages up to 2 GiB. It takes the messages out of the SCTP receive buffer and adds it to a separate reassembly buffer on the channel instance. However, there is no flow control signal because if that would get fed back to the receive buffer, 2) and 3) would suddenly become an issue.

**Comment by murillo128** [comment]
> For example, Firefox uses an SCTP receive buffer of 1 MiB but can receive messages up to 2 GiB. 

That's an implementation detail specific to Firefox, it could use 2 GB of arrive buffer and pass it zero copy to the JS layer.

**Comment by murillo128** [comment]
> For example, Firefox uses an SCTP receive buffer of 1 MiB but can receive messages up to 2 GiB.

What prevents Firefox for declaring a 2GiB buffer size on the first place?

**Comment by henbos** [comment]
If you empty the receive buffer and put it in a temporary message buffer then you can receive messages larger than the receive buffer, but having another buffer does not solve the congestion control problem. Eventually, the temporary message buffer will fill up and then either 1) you stop emptying the receive buffer to get congestion control to kick in, which is a variety of the existing proposal, or you 2) create more message buffers, in which case we have the same problem that we do now (it's just that the buffer is stored in "message buffers" instead of in in-queue PostTasks). If we are CPU-bound we have to have some sort of congestion control. Blocking messages from arriving sounds like a less severe problem than resource exhaustion.

**Comment by murillo128** [comment]
@henbos I agree, the `a_rwnd` buffer limit should not be there to protect the sctp stack but the whole browser. 

Having a nice looking 1MiB used in the sctp stack but 16GiB accumulating on other places of the browser doesn't seems be the correct usage of it.

**Comment by murillo128** [comment]
@lgrahl note that with the same mechanism (reducing the `a_rwnd` when the data is sent to the js app and not when passed from the sctp stack to the dc stack) would work when using your WATHWG  api proposal and you would not need any extra per-datachannel buffer size back pressure mechanism. 

You will decrease the `a_rwnd` when the `read` operation is done, so the only way of one datachannel draining the queue and stalling it is that the js app does not read from the datachannel (on purpose or due to a bug) . If any case, stalling is the correct behavior there IMO.

**Comment by lgrahl** [comment]
@murillo128 

> What prevents Firefox for declaring a 2GiB buffer size on the first place?

It would be an abuse of the receive buffer to work around an API issue. And it would break once someone activates ndata (which is a spec requirement). Needless to say that transport protocol implementations usually allocate the buffer at once... I think you can imagine what would happen. To quote RFC 4960:

>   Advertised Receiver Window Credit (a_rwnd): 32 bits (unsigned
>   integer)
>
>   This value represents the **dedicated buffer space**, in number of
>   bytes, the sender of the INIT has **reserved** in association with
>   this window.  During the life of the association, this buffer
>   space SHOULD NOT be lessened (i.e., dedicated buffers taken away
>   from this association); however, an endpoint MAY change the value
>   of a_rwnd it sends in SACK chunks.

Also, let's discuss stream API things in a separate issue.

---

@henbos

> If you empty the receive buffer and put it in a temporary message buffer then you can receive messages larger than the receive buffer, but having another buffer does not solve the congestion control problem.

It doesn't solve the problem. I just wanted to clarify that `receive buffer size != maximum message size` by an example.

> Eventually, the temporary message buffer will fill up and then either 1) you stop emptying the receive buffer to get congestion control to kick in, which is a variety of the existing proposal, or you 2) create more message buffers, in which case we have the same problem that we do now (it's just that the buffer is stored in "message buffers" instead of in in-queue PostTasks).

We can't do option 1) because as explained above (which is why Firefox doesn't). It would just expand the time period/message size but the issue would still occur eventually. 2) is, as you already said, not useful.

> If we are CPU-bound we have to have some sort of congestion control. Blocking messages from arriving sounds like a less severe problem than resource exhaustion.

Sure, but the current proposal has a much more severe effect than it's sole intention to *block individual messages* as I explained above.

Let's please call this *flow control* and not *congestion control*. :slightly_smiling_face: *Congestion control* is a different mechanism.

---

So, here's a suggestion what *could* be done which wouldn't result in all kinds of unwanted side effects: Don't feed decrypted packets from the DTLS stack to the SCTP stack when the event loop queue is congested. They will eventually be retransmitted if it's a reliable channel.

**Comment by murillo128** [comment]
> So, here's a suggestion what could be done which wouldn't result in all kinds of unwanted side effects: Don't feed decrypted packets from the DTLS stack to the SCTP stack when the event loop queue is congested. They will eventually be retransmitted if it's a reliable channel.

That will cause retransmissions and congestion on the wire possible causing disruption of the audio and video feed

**Comment by murillo128** [comment]
> It would be an abuse of the receive buffer to work around an API issue. 

Datachannel API is not well suited for SCTP, that's why we are here, so yes, we are already abusing it. 

> And it would break once someone activates ndata (which is a spec requirement).

That's true, but that's also true that when someone activates ndata and try to send 10 4GB files with current API (common torrent case), it will not work either with current API.

> Needless to say that transport protocol implementations usually allocate the buffer at once...

That's an implementation issue.

**Comment by lgrahl** [comment]
> Datachannel API is not well suited for SCTP, that's why we are here, so yes, we are already abusing it.

Your opinion.

> That will cause retransmissions and congestion on the wire possible causing disruption of the audio and video feed

Oh, so we suddenly started encrypting RTP packets with **DTLS**? :smirk: 

> That's true, but that's also true that when someone activates ndata and try to receive 10 4GB files with current API (common torrent case), it will not work either.

Do you want to compare spontaneous deadlocks with an explicit exception on a `.send` call?

**Comment by murillo128** [comment]
> Oh, so we suddenly started encrypting RTP packets with DTLS? 😏

Those SCTP chunks data retransmissions will use bandwidth, that won't be available for audio and video. Note that audio/video may be hardware accelerated and not affected by the js loop at all.

> Do you want to compare spontaneous deadlocks with an explicit exception on a .send call?

I am speaking about receiving data, not sure how the .send call is involved. And I was referring that without a new API we will be in this situation anyway. Maybe the sctp transmission will not be halted, but the browser will die when having reserved several GBs on intermediary buffers. So yes, I prefer the sctp spontaneous deadlock (Note: I am speaking without an API change).

**Comment by lgrahl** [comment]
> Those SCTP chunks data retransmissions will use bandwidth, that won't be available for audio and video. Note that audio/video may be hardware accelerated and not affected by the js loop at all.

Ah, I misread your comment. That is a different story though. We have the competition discussion longer than I've been participating and I doubt it's going to make much of a difference if we intentionally drop packets because the event loop queue is already overburdened. The browser will likely have different issues once it reaches that point.

> I am speaking about receiving data, not sure how the .send call is involved.

`.send` raises an exception if one tries to send a message larger than the other peer supports.

> And I was referring that without a new API we will be in this situation anyway.

What situation are we in with the current API?
Can spontaneously deadlock? No.
Can go OOM? Perhaps in dire circumstances. I wouldn't judge browsers killing data channels or even peer connections that go insane with memory usage as long as it's documented.

> Maybe the sctp transmission will not be halted, but the browser will die when having reserved several GBs on intermediary buffers. So yes, I prefer the sctp spontaneous deadlock (Note: I am speaking without an API change).

Then let's agree that we will not agree on this.

**Comment by murillo128** [comment]
> We have the competition discussion longer than I've been participating and I doubt it's going to make much of a difference if we intentionally drop packets because the event loop queue is already overburdened. The browser will likely have different issues once it reaches that point.

One thing is competition, and another one is causing unneeded retransmisión.  The RTP loop is not going in the main js thread, also audio/video encoding/decoding can be done in hardware, so media can be perfectly fine even if js loop is causing cpu overuse, specially on low end devices with dedicated media chips.

>.send raises an exception if one tries to send a message larger than the other peer supports.

I was only referring to the receiving side (sender can be a server with proper streaming api implemented). 

> Can spontaneously deadlock? No.
Can go OOM? Perhaps in dire circumstances. I wouldn't judge browsers killing data channels or even peer connections that go insane with memory usage as long as it's documented.

Ok, so if we instead of causing a deadlock, we document that in case that the `a_wnd` is exhausted and the browser has no completed messages to deliver to the api, then it will close the datachannel/discard the current messages/close pc, would it be ok then?

**Comment by lgrahl** [comment]
> One thing is competition, and another one is causing unneeded retransmisión.

Fair enough. But like I said, it's a workaround.

> I was only referring to the receiving side (sender can be a server with proper streaming api implemented).

I think we're talking about two different things.

> Ok, so if we instead of causing a deadlock, we document that in case that the `a_wnd` is exhausted and the browser has no completed messages to deliver to the api, then it will close the datachannel/discard the current messages/close pc, would it be ok then?

No, of course that's not acceptable.

**Comment by murillo128** [comment]
> >One thing is competition, and another one is causing unneeded retransmisión.

> Fair enough. But like I said, it's a workaround.

Using someone else words: No, of course that's not acceptable to degrade audio and video with unneeded sctp retransmissions.. 

> > I was only referring to the receiving side (sender can be a server with proper streaming api implemented).

> I think we're talking about two different things.

Could you then comment on mine?

**Comment by lgrahl** [comment]
> Could you then comment on mine?

I'll go back to the original statement:

> [...] when someone activates ndata and try to receive 10 4GB files with current API (common torrent case), it will not work either.

When calling `.send`, you're also restricted by the remote peer's announced maximum message size. No one announces 4 GiB at the moment. That was the direction my thoughts went.

But let's ignore `.send` because you wanted to talk about the receiving side. Assume A uses Firefox (which supports receiving up to 2 GiB sized messages) and B is a server that has a streaming API. Let's also assume they have activated ndata. Then that server could effectively start sending 2 GiB sized messages on every channel with the same priority. That would overburden A and bring it into a OOM situation.

Okay, granted. :ok_hand: Someone may do that at some point. But it's a considerable more unlikely scenario compared to receiving a message larger than the receive buffer - with and without ndata. That already happens when you send something > 1 MiB to Firefox. In Firefox < 64 you already surpass the receive buffer with just slightly more than 128 KiB. I don't believe the likelihood of these two occurrences are in the same league. And if one adjusts the receive buffer of Firefox to 2 GiB, it'd run into the same issue again once ndata is being deployed. That would be a spec conflict anyway since ndata is mandatory.

---

So far we've established three possibilities:

1. Don't clear the receive buffer before having handed out a message. Requires `max-message-size` to be locked to the receive buffer's size, otherwise instant deadlock if the receive buffer is smaller than the message received. Risk spontaneous deadlocks if ndata (which is mandatory) is activated.
2. Drop decrypted DTLS packets intended for the SCTP stack when the event loop is congested. Risk retransmissions leading to higher bandwidth usage and potential media quality degradation.
3. Do nothing and declare the current API unsuitable. Risk the receiving peer to go OOM (or let the browser kill the channel if it reaches an absurd amount of memory usage).

Pick your poison. These are my comments for each possibility:

1. No (said enough about why already).
2. Can't estimate the risk here but I don't expect the event queue to be congested that often, so the impact may be acceptable. However, that is only my guess. I don't know how often the event queue would be congested. And it would require a definition of when the event queue is considered *congested*.
3. From my own experience, I haven't heard complains about this issue so far. I'm also heavily using data channels for various projects and haven't experienced it to be an issue in practice. Perhaps because usrsctp is usually much earlier CPU bound than the event queue and the JS thread. User applications can also throttle themselves. Sure, it's not great but maybe this problem is slightly overrated? I'd vote for this.

**Comment by lgrahl** [comment]
> Can we also agree that given that the maximum value supported by SCTP for a_rwnd is 4Gib, choosing a 1MiB value in Firefox is an arbitrary decision (probably driven by the sctp stack used) and that nothing prevents other datachannels implementations to use much higher values and implement it as described? 

If other data channel implementations would implement it as described (I'm assuming you're referring to @henbos proposal here), they would also need to implement ndata in which case it's prone to deadlock. And with or without ndata it may have an impact on throughput as well if we keep the receive buffer filled a lot of the time.

> [...] But there is a 4) option, that mixes 1), considering current API as unsuitable and your stream API. Please, let me explain [...]

Let's put the ndata issue aside for a second (and that I oppose to the changes to tie the maximum message size to the receive buffer) because I can keep repeating that forever. But don't forget that this is a real problem that would need to be sorted out and, frankly, I don't know how that could be done at this point.

I think in theory your idea should work. The PPID idea is worth exploring regardless of the outcome of this issue as it may be a necessity for allowing us to ignore `max-message-size`. On the flip side it requires clear commitment from the application to use only one of the APIs.

**Comment by henbos** [comment]
If you don't think being CPU-bound is a big issue then you shouldn't be affected by CPU-bound applications letting the receive buffer get congested, as long as it doesn't have significant downsides for non-CPU-bound scenarios.

I think we can allow [larger messages than receive buffers](http://i.imgur.com/4lpVkTS.gif) with intermediate buffers, and perhaps we could avoid data channels from monopolizing by letting the sender allocate messages for its channels before other channels are allowed to steal that buffer space. If [stream schedulers](https://tools.ietf.org/html/draft-ietf-tsvwg-sctp-ndata-13#section-3) like a round robin scheduler (SCTP_SS_RR) with interleaving enabled is used we could prevent starvation too?

Forgive me if all of this has been covered before or if I'm speaking out of ignorance. I am not an SCTP expert. I'm throwing ideas out there with little background.

**Proposal Part 1: The receiver**

> Let each underlying SCTP data channel have an internal slot [[MessageBuffer]] that is a buffer of size max-message-size. (Note: We will only touch [[MessageBuffer]] on the network thread.)
>
> **process chunk steps** (on network thread): When receiving an I-DATA chunk into the receive buffer a_rwnd is decremented by the size of the chunk. If there is enough space in the corresponding channel's [[MessageBuffer]], move the chunk into [[MessageBuffer]] and increment a_rwnd by the size of the chunk.
> When a [[MessageBuffer]] is full, queue a task (to the JavaScript thread) with a _message_ set to the value of [[MessageBuffer]] that runs the following steps:
> 1. In parallel, run the following steps (on the network thread):
> - Clear [[MessageBuffer]] and run the **process chunk steps** on any I-DATA chunks for this channel that is currently in the receive buffer.
> 2. Convert _message_ to the appropriate JavaScript object and fire "onmessage" with it.

This takes care of the receiver side without sacrificing large messages. One data channel can block other data channels if that channel's buffer is full _AND_ the receive buffer is full. But we would only get this far if we are CPU-bound and can't handle more messages anyway. As long as chunks are arriving for all channels, all channels will have their [[MessageBuffer]] filled up and emptied accordingly. You _can't_ ignore a particular channel, all channels' messages are handled (whether or not the application has an onmessage listener).

If the receive buffer gets full, whether because of network performance or being CPU-bound, the sender will know about it with `rwnd`.

**Proposal Part 2: The sender**

We want to avoid one channel monopolizing the send buffer. (E.g. channel A sends lots of small messages so that channel B cannot send its large messages because the buffer is full. As soon as the buffer frees up a little bit, channel A fills it up again. Channel B is a sad data channel.) These are examples of `rwnd` being low or messages being really big.

New API:
```
partial interface RTCSctpTransport {
  Promise<unsigned long> allocateSendMessage(unsigned long size);
};
```

_allocateSendMessage()_ will affectively decrease the _available buffer size of the send()_ (see step 5 of [send()](https://w3c.github.io/webrtc-pc/#dfn-send)). When the true underlying buffer has _size_ amount of space available, the _available buffer size of the send()_ will be increased by _size_ again and the return promise is resolved with _size_. This guarantees that subsequent send() calls with messages of this size will succeed step 5, and in the meantime blocking other channels from performing send() if that would cause our allocated message to not fit into the buffer.

Multiple allocations (multiple allocateSendMessage() calls) are queued. Example:
```
sctpTransport.allocateSendMessage(aMessage.length).then(() => a.send(aMessage));
sctpTransport.allocateSendMessage(bMessage.length).then(() => b.send(bMessage));
```
If both messages fit in the buffer at the same time both promises will be resolved immediately, otherwise a's allocation is queued before b's allocation. There is no way for b's message(s) to block a's message from sending because of the order. This allows custom application logic for how to prioritize different channels, e.g. "for every X message/bytes sent by channel A I want Y messages/bytes to be sent by channel B".

Does this address most of the issues within the scope of 1.0, or is are my proposals fundamentally flawed?

**Comment by henbos** [comment]
@aboba Can you be the devil's advocate for the proposal in [my previous comment](https://github.com/w3c/webrtc-pc/issues/2086#issuecomment-458724853)?

**Comment by murillo128** [comment]
BTW, are there any w3c guidelines regarding how APIs should manage/prevent out of memory, resource exhaustion or cpu overuse or it is something left out to each implementation?

**Comment by lgrahl** [comment]
@henbos

> If stream schedulers like a round robin scheduler (SCTP_SS_RR) with interleaving enabled is used we could prevent starvation too?

That is my understanding of it. :slightly_smiling_face:

To your proposal: 

#### Receiver side (flow control signal)

If `[[MessageBuffer]]` is lazily allocated then this might be feasible unless I'm missing something. However, it does require that the sending peer actually obeys `max-message-size` as designated by the receiving peer (I know a popular browser that still doesn't \*cough cough\*.), so backwards compatibility is something that would need to be discussed.

#### Sender side (stream/channel monopolisation)

I don't think this needs to be addressed: Analogous to having a separate message receive buffer on each channel, an implementation can have a message send buffer for each channel and do round-robin scheduling of different channel's chunks when putting them into the SCTP send buffer. [Firefox does exactly that](https://dxr.mozilla.org/mozilla-central/rev/c2593a3058afdfeaac5c990e18794ee8257afe99/netwerk/sctp/datachannel/DataChannel.cpp#1261). It doesn't fix it completely since whoever comes first still initially fills the send buffer but as soon as a little bit of space becomes available, you're good. And sure, you would still eventually need to throw `OperationError` if the accumulated buffer sizes are becoming an issue but my understanding is that this is a last resort mechanism in case an application ignores *unacceptably high* `bufferedAmount` values (I'll come back to the definition of *unacceptably high* later).

Furthermore, the concrete proposal seems a bit brittle to me, especially when it comes to string messages. Determining the byte length of a string message would be easy to get wrong. It would also be a late API change and (hopefully) obsolete with the streams extension.

This is more of a side note: Adding an attribute to determine the total available buffer space is probably cheap and would be very helpful to determine what a *unacceptably high* is. At the moment, the application has to guess and trigger an exception to determine it.

**Comment by murillo128** [comment]
> This is more of a side note: Adding an attribute to determine the total available buffer space is probably cheap and would be very helpful to determine what unacceptably high is. At the moment, the application has to guess and trigger an exception to determine it.

+1

**Comment by alvestrand** [comment]
Count my vote for making #1732 real and not adding more hacks to the existing API.
